### ISSUE TITLE
Distinguish Identifier from IdentifierPattern.

### ISSUE THREAD

**Issue by aaditmshah** [estree/estree#196]
This would make it easier to calculate the free variables of a program. Every time we see an `Identifier` would we add it to the list of free variables, and every time we see an `IdentifierPattern` we would add it to the list of free patterns. This distinction is important because of `AssignmentPattern`, which is defined as follows:

```javascript
interface AssignmentPattern <: Pattern {
    type: "AssignmentPattern";
    left: Pattern;
    right: Expression;
}
```

As you can see, an `AssignmentPattern` can have both free variables and free patterns. An example would be `(a=b) => {}`, where the free variable is `b` and the free pattern is `a`. By conflating `Identifier` and `IdentifierPattern`, it becomes difficult to find the free variables in a program because an `Identifier` could be a variable reference or it could be a pattern variable, depending upon the context.

Since ESTree aims to be contextless, distinguishing an `Identifier` from an `IdentifierPattern` would make a lot of sense. Without this distinction, we need to be aware of the context of an identifier whenever we encounter it.

I am aware that suggested modifications to the specification should be backwards compatible, and that this suggestion is not backwards compatible. Hence, a more reasonable solution would be to extend the `Identifier` interface as follows:

```javascript
extend interface Identifier {
    pattern: boolean;
}
```

**Comment by not-an-aardvark** [comment]
This would be very useful. In ESLint we have a lot of places where we have to resolve this manually, and it's not completely trivial to do so. If backwards compatibility weren't a concern I would want to make these two separate node types (as e.g. the Shift AST does), but adding a property as a workaround seems worthwhile to me.

**Comment by adit-hotstar** [comment]
I solved this problem by defining a fold over the ESTree data type which adds a synthetic `isPattern` property to the `Identifier` and `MemberExpression` interfaces.

```javascript
const foldNode = algebra => function loop(node, isPattern=false) {
    const pattern = node => loop(node, true);
    const recurse = node => loop(node, false);
    const recur = isPattern ? pattern : recurse;
    const type = node ? node.type : "Default";
    const fold = algebra[type];

    switch (fold && type) {
    case "Identifier":               return fold(node.name, isPattern);
    case "Literal":                  return fold(node.value, node.regex);
    case "Program":                  return fold(node.body.map(recurse), node.sourceType);
    case "ExpressionStatement":      return fold(recurse(node.expression), node.directive);
    case "BlockStatement":           return fold(node.body.map(recurse));
    case "EmptyStatement":           return fold;
    case "DebuggerStatement":        return fold;
    case "WithStatement":            return fold(recurse(node.object), recurse(node.body));
    case "ReturnStatement":          return fold(recurse(node.argument));
    case "LabeledStatement":         return fold(recurse(node.label), recurse(node.body));
    case "BreakStatement":           return fold(recurse(node.label));
    case "ContinueStatement":        return fold(recurse(node.label));
    case "YieldExpression":          return fold(recurse(node.argument), node.delegate);
    case "IfStatement":              return fold(recurse(node.test), recurse(node.consequent), recurse(node.alternate));
    case "SwitchStatement":          return fold(recurse(node.discriminant), node.cases.map(recurse));
    case "SwitchCase":               return fold(recurse(node.test), node.consequent.map(recurse));
    case "ThrowStatement":           return fold(recurse(node.argument));
    case "TryStatement":             return fold(recurse(node.block), recurse(node.handler), recurse(node.finalizer));
    case "CatchClause":              return fold(pattern(node.param), recurse(node.body));
    case "WhileStatement":           return fold(recurse(node.test), recurse(node.body));
    case "DoWhileStatement":         return fold(recurse(node.body), recurse(node.test));
    case "ForStatement":             return fold(recurse(node.init), recurse(node.test), recurse(node.update), recurse(node.body));
    case "ForInStatement":           return fold(pattern(node.left), recurse(node.right), recurse(node.body));
    case "ForOfStatement":           return fold(pattern(node.left), recurse(node.right), recurse(node.body), node.await);
    case "FunctionDeclaration":      return fold(pattern(node.id), node.params.map(pattern), recurse(node.body), node.async, node.generator);
    case "VariableDeclaration":      return fold(node.kind, node.declarations.map(recurse));
    case "VariableDeclarator":       return fold(pattern(node.id), recurse(node.init));
    case "ThisExpression":           return fold;
    case "ArrayExpression":          return fold(node.elements.map(recurse));
    case "ObjectExpression":         return fold(node.properties.map(recurse));
    case "Property":                 return fold(recurse(node.key), recur(node.value), {kind: node.kind, method: node.method, shorthand: node.shorthand, computed: node.computed});
    case "FunctionExpression":       return fold(pattern(node.id), node.params.map(pattern), recurse(node.body), node.async, node.generator);
    case "ArrowFunctionExpression":  return fold(pattern(node.id), node.params.map(pattern), recurse(node.body), node.expression, node.async, node.generator);
    case "UnaryExpression":          return fold(node.operator, recurse(node.argument), node.prefix);
    case "UpdateExpression":         return fold(node.operator, recurse(node.argument), node.prefix);
    case "BinaryExpression":         return fold(node.operator, recurse(node.left), recurse(node.right));
    case "AssignmentExpression":     return fold(node.operator, pattern(node.left), recurse(node.right));
    case "LogicalExpression":        return fold(node.operator, recurse(node.left), recurse(node.right));
    case "MemberExpression":         return fold(node.computed, recurse(node.object), recurse(node.property), isPattern);
    case "ConditionalExpression":    return fold(recurse(node.test), recurse(node.consequent), recurse(node.alternate));
    case "CallExpression":           return fold(recurse(node.callee), node.arguments.map(recurse));
    case "NewExpression":            return fold(recurse(node.callee), node.arguments.map(recurse));
    case "SequenceExpression":       return fold(node.expressions.map(recurse));
    case "Super":                    return fold;
    case "SpreadElement":            return fold(recurse(node.argument));
    case "TemplateLiteral":          return fold(node.quasis.map(recurse), node.expressions.map(recurse));
    case "TaggedTemplateExpression": return fold(recurse(node.tag), recurse(node.quasi));
    case "TemplateElement":          return fold(node.value, node.tail);
    case "ObjectPattern":            return fold(node.properties.map(pattern));
    case "ArrayPattern":             return fold(node.elements.map(pattern));
    case "RestElement":              return fold(pattern(node.argument));
    case "AssignmentPattern":        return fold(pattern(node.left), recurse(node.right));
    case "ClassDeclaration":         return fold(pattern(node.id), recurse(node.superClass), recurse(node.body));
    case "ClassExpression":          return fold(pattern(node.id), recurse(node.superClass), recurse(node.body));
    case "ClassBody":                return fold(node.body.map(recurse));
    case "MethodDefinition":         return fold(recurse(node.key), recurse(node.value), {kind: node.kind, computed: node.computed, static: node.static});
    case "MetaProperty":             return fold(recurse(node.meta), recurse(node.property));
    case "ImportDeclaration":        return fold(node.specifiers.map(recurse), recurse(node.source));
    case "ImportSpecifier":          return fold(recurse(node.imported), recurse(node.local));
    case "ImportDefaultSpecifier":   return fold(recurse(node.local));
    case "ImportNamespaceSpecifier": return fold(recurse(node.local));
    case "ExportNamedDeclaration":   return fold(recurse(node.declaration), node.specifiers.map(recurse), recurse(node.source));
    case "ExportSpecifier":          return fold(recurse(node.local), recurse(node.exported));
    case "ExportDefaultDeclaration": return fold(recurse(node.declaration));
    case "ExportAllDeclaration":     return fold(recurse(node.source));
    case "AwaitExpression":          return fold(recurse(node.argument));
    case "Default":                  return fold;
    default:                         throw new Error(`Can't handle node ${type}`);
    }
};
```

**Comment by adit-hotstar** [comment]
@RReverser @dherman @michaelficarra Any update on this issue?

**Comment by RReverser** [comment]
Aside from backwards compatibility you mentioned, another item from Philisophy section declared in the README is

> Contextless: Nodes should not retain any information about their parent. ie. a FunctionExpression should not be aware of if it's a concise method.

What you're suggesting seems to directly contradict that rule and encode the context in the node:

> Since ESTree aims to be contextless, distinguishing an Identifier from an IdentifierPattern would make a lot of sense. Without this distinction, we need to be aware of the context of an identifier whenever we encounter it.

That is, ESTree AST being contextless means that information that can be deduced from the context should not be additionally encoded/duplicated in the node itself, and not that consumers can avoid tracking the context.

**Comment by adit-hotstar** [comment]
@RReverser 

> That is, ESTree AST being contextless means that information that can be deduced from the context should not be additionally encoded/duplicated in the node itself, and not that consumers can avoid tracking the context.

By that logic, the current ESTree AST isn't contextless either because:

1. It distinguishes [`ObjectExpression`](https://github.com/estree/estree/blob/master/es5.md#objectexpression) and [`ObjectPattern`](https://github.com/estree/estree/blob/master/es2015.md#objectpattern).
2. It distinguishes [`ArrayExpression`](https://github.com/estree/estree/blob/master/es5.md#arrayexpression) and [`ArrayPattern`](https://github.com/estree/estree/blob/master/es2015.md#arraypattern).

Let's compare the definitions of `ArrayExpression` and `ArrayPattern`.

```webidl
interface ArrayExpression <: Expression {
    type: "ArrayExpression";
    elements: [ Expression | null ];
}

interface ArrayPattern <: Pattern {
    type: "ArrayPattern";
    elements: [ Pattern | null ];
}
```

Now, consider the definition of [`Identifier`](https://github.com/estree/estree/blob/master/es5.md#identifier).

```webidl
interface Identifier <: Expression, Pattern {
    type: "Identifier";
    name: string;
}
```

Why this double standard? By your definition of &ldquo;contextless&rdquo; you should have a single `Array` interface instead of `ArrayExpression` and `ArrayPattern`. After all, both `ArrayExpression` and `ArrayPattern` are &ldquo;retaining information about their parent&rdquo;.

```webidl
interface Array <: Expression, Pattern {
    type: "Array";
    elements: [ Expression | Pattern | null ];
}
```

Now have fun figuring out whether `Array` is actually an `ArrayExpression` or an `ArrayPattern`. Clearly, it's not the concern of the AST to distinguish this because it needs to be &ldquo;contextless&rdquo;. Why duplicate information that be deduced from context? It's the job of the consumer to track context.

---

All I'm asking is that you allow consumers to distinguish whether `Identifier` is actually an `IdentifierExpression` or an `IdentifierPattern` because you have conflated the two.

```webidl
interface IdentifierExpression <: Expression {
    type: "IdentifierExpression";
    name: string;
}

interface IdentifierPattern <: Pattern {
    type: "IdentifierPattern";
    name: string;
}
```

You're already doing this for `ObjectExpression` and `ObjectPattern`, and `ArrayExpression` and `ArrayPattern`. So, why should `Identifier` get the short end of the stick?

Now, I understand that we need to maintain backward compatibility. That's the reason why I proposed the following backward compatible extension to the `Identifier` interface.

```webidl
extend interface Identifier {
    pattern: boolean;
}
```

---

Finally, let's talk about being &ldquo;contextless&rdquo; because I think it's important to clarify what that means.

> **Contextless:** Nodes should not retain any information about their parent. ie. a `FunctionExpression` should not be aware of if it's a concise method. (eg. #5)

Clearly, a `FunctionExpression` has no business knowing whether or not it's a concise method. That information should belong to the parent `Property` node.

However, it does make sense for an `Identifier` to know whether it's an expression or a pattern. That is information about `Identifier` itself. It is not information about any parent node of `Identifier`.

Sure, you can also deduce this information based on context. However, that doesn't mean that my proposition is not contextless, because the information pertains to the `Identifier` itself.

---

**TLDR:**

> Nodes should not retain any information about their parent.

My proposition does not make `Identifier` retain any information about its parent. The new information pertains to `Identifier` itself.

If you believe otherwise, then by your own logic you must also believe that `ObjectExpression`, `ObjectPattern`, `ArrayExpression`, and `ArrayPattern` are not contextless.

**Comment by RReverser** [comment]
> By that logic, the current ESTree AST isn't contextless either because:

You are right about similarily, but it's the reasons why separate nodes exist that matter here.

The reason these types are separate isn't to encode the outer context they're in, but because their inner contents are different.

`ObjectExpression` / `ArrayExpression` can contain arbitrary expression nodes, whereas `ObjectPattern` / `ArrayPattern` can contain only nodes that adhere to pattern rules - for example, you can't put a `CallExpression` or `BinaryExpression` inside pattern nodes, but you can in expression nodes.

In `Identifier` case the structure and contents are the same, no matter whether it's a pattern or an expression.

**Comment by RReverser** [comment]
> That is information about Identifier itself. It is not information about any parent node of Identifier.

Is it though? It's not uncommon for ESTree consumers to take identifier from a pattern position and move or copy to an expression position (e.g. to transform `let {a, b} = obj` into `let a = obj.a, b = obj.b;` while preserving all location info and so on).

Your proposed change would break this use-case and require updating existing code that relies on this, thus additionally breaking

> Backwards compatible: Non-additive modifications to existing constructs will not be considered unless immense support is in favor of such changes. (eg. #65)

**Comment by adit-hotstar** [comment]
@RReverser

> The reason these types are separate isn't to encode the outer context they're in, but because their inner contents are different.

How does that matter? The inner contents are entirely determined by the outer context. If the outer context is an expression then the inner contents can only be `Expression` nodes. On the other hand, if the outer context is a pattern then the inner contents can only be `Pattern` nodes.

For example, you can't put a `CallExpression` or a `BinaryExpression` inside a pattern context but you can put them inside an expression context. So why duplicate information that can be deduced from context? It's the job of the producer to create a valid AST based on the context.

Also, it doesn't matter why you're distinguishing `ArrayExpression` from `ArrayPattern`. By distinguishing them, you are going against your own definition of “contextless” nodes. In addition, as I just demonstrated you don't need to distinguish them at all just for the sake of the producer.

Just as it is the job of the consumer to keep track of the context to determine whether a node is an expression or a pattern, similarly it is the job of the producer to keep track of the context to determine whether a node can contain expressions or patterns.

```webidl
//           this is a problem for consumers
//                          |
//                 +--------+--------+
//                 |                 |
interface Array <: Expression, Pattern {
    type: "Array";
    elements: [ Expression | Pattern | null ];
} //            |                  |
//              +---------+--------+
//                        |
//         this is a problem for producers
```

Clearly, you can see that your own definition of “contextless” nodes is a problem for producers. However, you are conveniently ignoring the fact that it is also a problem for consumers. Why this double standard? Why do consumers get the short end of the stick?

---

Now, to elucidate why this is a problem for consumers consider the definitions of [`FunctionDeclaration`](https://github.com/estree/estree/blob/master/es5.md#functiondeclaration), [`FunctionExpression`](https://github.com/estree/estree/blob/master/es5.md#functionexpression), [`ClassDeclaration`](https://github.com/estree/estree/blob/master/es2015.md#classdeclaration), and [`ClassExpression`](https://github.com/estree/estree/blob/master/es2015.md#classexpression).

```webidl
interface FunctionDeclaration <: Function, Declaration {
    type: "FunctionDeclaration";
    id: Identifier;
}

interface FunctionExpression <: Function, Expression {
    type: "FunctionExpression";
}

interface ClassDeclaration <: Class, Declaration {
    type: "ClassDeclaration";
    id: Identifier;
}

interface ClassExpression <: Class, Expression {
    type: "ClassExpression";
}
```

Each of these interfaces inherits from two other interfaces.

1. A `FunctionDeclaration` is a `Function` and a `Declaration`.
2. A `FunctionExpression` is a `Function` and an `Expression`.
3. A `ClassDeclaration` is a `Class` and a `Declaration`.
4. A `ClassExpression` is a `Class` and an `Expression`.

Now, compare this to the definition of the [`Identifier`](https://github.com/estree/estree/blob/master/es5.md#identifier) interface.

```webidl
interface Identifier <: Expression, Pattern {
    type: "Identifier";
    name: string;
}
```

Do you see the blunder that you made?

* What you're saying is that an `Identifier` is an `Expression` and a `Pattern`.
* However, what you meant to say is that an `Identifier` is either an `Expression` or a `Pattern`.

This has nothing to do with the “the structure and contents” of `IdentifierExpression` and `IdentifierPattern` being the same. You made a blunder. Plain and simple. You used an `and` operator instead of using an `or` operator, and thereby changed the semantics of the AST.

The only reason you didn't notice this problem sooner is because `Expression` and `Pattern` are both empty interfaces. If they weren't empty then you would have noticed the problem immediately, because every `Identifier` in expression position would be required to implement the `Pattern` interface and vice versa.

What's worse is that you made this blunder twice. The second time with [`MemberExpression`](https://github.com/estree/estree/blob/master/es5.md#memberexpression).

```webidl
interface MemberExpression <: Expression, Pattern {
    type: "MemberExpression";
    object: Expression;
    property: Expression;
    computed: boolean;
}
```

---

> > That is information about `Identifier` itself. It is not information about any parent node of `Identifier`.
> 
> Is it though?

Yes, it is. Furthermore, your counterexample of taking an identifier in pattern position and moving it to expression position doesn't disprove my statement. The information of whether an `Identifier` is an expression or a pattern still pertains to the `Identifier` itself. It doesn't pertain to any parent node of the `Identifier`.

As for moving the identifier from pattern position to expression position, simply don't mutate the original data structure. You will still be able to transform `let {a, b} = obj` into `let a = obj.a, b = obj.b` while preserving all location info and so on.

```javascript
const toIdentifierExpression = ({ name, loc }) => ({
    type: "Identifier",
    pattern: false,
    name,
    loc
});

const toIdentifierPattern = ({ name, loc }) => ({
    type: "Identifier",
    pattern: true,
    name,
    loc
});
```

> Your proposed change would break this use-case and require updating existing code that relies on this, thus additionally breaking

No, it wouldn't. A transform which isn't aware of the `pattern` attribute would simply ignore it. It might transform the code into an invalid AST, but it wouldn't break anything. At worse, it will set `pattern` to `true` for an `Identifier` in expression position or vice versa. That is something that can be automatically fixed by linting tools. Simply correct the information if it doesn't match the context.

Furthermore, code generators wouldn't even care about the `pattern` attribute. The generated code is exactly the same for `Identifier` nodes in both expression position and pattern position. Hence, code generators wouldn't even need to be updated.

>> **Backwards compatible:** Non-additive modifications to existing constructs will not be considered unless immense support is in favor of such changes. (eg. #65)

My proposed solution is strictly additive. It doesn't subtract anything from the current specification. Hence, it's 100% backwards compatible.

**Comment by RReverser** [comment]
> However, you are conveniently ignoring the fact that it is also a problem for consumers.

> Do you see the blunder that you made?

Content of the discussion aside, I'd like to remind that this is an OSS project led by volunteers and so would ask to refrain from negative comments and personal accusations.

**Comment by adit-hotstar** [comment]
@RReverser I am sorry. I tend to get carried away some times. I'll be more respectful.

**Comment by adit-hotstar** [comment]
Also, thank you for engaging me in discussion. I know that you aren't obliged to, and I am glad that you are hearing me out.

I would like for something fruitful to come from our discussions so that it isn't a waste our time. I firmly believe that my proposal would make the ESTree specification better.

When I raised this issue 8 months ago, I did make sure that my proposed solution followed the philosophy of ESTree. I still do believe so, and I hope to convince you of the same.

**Comment by RReverser** [comment]
Sorry, I'm going to close this issue for the reasons outlined earlier. This information is still clearly contextual and, as you mentioned yourself, this would be a breaking change to change types. As for the additive variant:

```ts
extend interface Identifier {
    pattern: boolean;
}
```

This is something that tools can extend `Identifier` with themselves, since it doesn't conflict with any ESTree-defined properties, and it's common practice to attach extra temporary information to the AST nodes that way.

As such, I'm closing this issue in favour of userland implementation of specific tooling that needs to temporarily store this contextual information.
